### PR TITLE
add uploadPasswordCmd to read hackage password from arbitrary program output

### PR DESCRIPTION
**Pull request by strake**

### CODE REVIEW

File: cabal-install/Distribution/Client/Setup.hs [hunk 1]
```diff
@@ -1444,18 +1444,20 @@ instance Monoid InstallFlags where
...
 
 data UploadFlags = UploadFlags {
-    uploadCheck     :: Flag Bool,
-    uploadUsername  :: Flag Username,
-    uploadPassword  :: Flag Password,
-    uploadVerbosity :: Flag Verbosity
+    uploadCheck       :: Flag Bool,
+    uploadUsername    :: Flag Username,
+    uploadPassword    :: Flag Password,
+    uploadPasswordCmd :: Flag [[Char]],
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/Setup.hs]: Why not `Flag [String]`?

File: cabal-install/Main.hs [hunk 2]
```diff
@@ -952,12 +955,15 @@ uploadAction uploadFlags extraArgs globalFlags = do
       globalFlags' = savedGlobalFlags config `mappend` globalFlags
       tarfiles     = extraArgs
   checkTarFiles extraArgs
+  maybe_password <- case uploadPasswordCmd uploadFlags'
+                    of Flag (xs:xss) -> Just . Password <$> getProgramInvocationOutput verbosity (simpleProgramInvocation xs xss)
```
Comment (23Skidoo) [inline cabal-install/Main.hs]: Please keep all lines <80 col long.

**Comment by 23Skidoo** [comment]
This fails on Travis because older GHCs don't have `<$>` et al. in the Prelude.

File: cabal-install/Distribution/Client/Setup.hs (hunk 1 above)
Comment (strake) [inline cabal-install/Distribution/Client/Setup.hs]: I normally use [Char]. Cabal team prefers String?

Comment (23Skidoo) [inline cabal-install/Distribution/Client/Setup.hs]: Yes, `[String]` is preferable.

**Comment by 23Skidoo** [comment]
Still fails on Travis. You should put `import Control.Applicative (pure, (<$>))` inside the `#if !MIN_VERSION_base(4,8,0)` block.

**Comment by 23Skidoo** [comment]
Otherwise LGTM.

**Comment by 23Skidoo** [comment]
Merged, thanks!

**Pull request merged**